### DISCUSSION TITLE
Community Suggestions and Requests

### DISCUSSION THREAD

**Discussion by sbordyak** [CIRDLES/Tripoli#34] [Ideas]
Please share any suggestions or requests:
- Feature Requests - What should Tripoli be able to do?
- What (meta)data would you like to be displayed and how should they be displayed?
- Lab workflows - How can Tripoli best integrate with your lab?
- What’s missing?

**Comment by bowring** [comment]
Drag and drop of raw data files

**Comment by noahmclean** [comment]
Maria: Combine data files from different methods that measure the same isotope ratios on the same sample.

**Reply by noahmclean** [reply]
This will require independent fractionation correction for each method if they're on different collectors.

**Comment by ryanickert** [comment]
More flexibility in the formatting of output "summarized" data.  On OG Tripoli I'm referring to what you get by clicking "Copy all means with pct error".

**Reply by noahmclean** [reply]
Great idea!  Next we need to come up with a list of what these outputs look like.  Can you provide some examples for what an exported (csv?) file might look like?

**Reply by ryanickert** [reply]
Yes!

For example, a single row with no headers with each result accompanied by absolute uncertainty, then after that the correlation coefficients:

Example output:

18.51561, 0.01235, 15.43867, 0.01565, 37.54687, 0.026578, 0.953, 0.948, 0.965

**Reply by ryanickert** [reply]
From Pieter: "IsoplotR will soon allow other programs to populate it with externally generated json files. It would be nice if Tripoli would include a button to take advantage of this function.:

**Comment by ryanickert** [comment]
Permit negative values please!

**Comment by noahmclean** [comment]
David: add something simple to the software so that you can track all the corrections and the uncertainty that resulted from them.

**Comment by ryanickert** [comment]
I would like the ability to write user functions that can be saved and then applied in Tripoli.  Ideally there would be functionality to refer to measurements from different cycles so that I can do something like calculate a beam decay/growth rate parameter.

**Reply by bowring** [reply]
example functions?

**Reply by ryanickert** [reply]
slope = (i_t2-i_t1)/(t2-t1)

Where i_t1 and i_t2 are an ion beam intensity at time 1 and time 2, and t1 and t2 are time 1 and time 2.

**Reply by ryanickert** [reply]
R64 = (R65_m - R65_t)-(R45_m - R45_t)

RXY is a ratio of X/Y
_m is the measured ratio
_t is the spike isotopic composition (effectively a constant external to the analysis)

**Reply by bowring** [reply]
did you check out Squid3 expressions / video?

**Reply by ryanickert** [reply]
hm no

**Reply by ryanickert** [reply]
Ok just took a quick look.  It looks very complicated!

**Reply by schoeneblair** [reply]
I agree with this one, to have more options for interference corrections, fractionation corrections, etc.

**Comment by noahmclean** [comment]
PietR: need a common format to communicate with IsoplotR, so that with one click we can move data along with its traceability.

**Reply by pvermees** [reply]
PietR :laughing:

**Comment by pvermees** [comment]
It would be nice if Tripoli came with an API that would allow advanced users to call individual functions from the command line, or create automation scripts.

**Comment by dmcgee314** [comment]
It would be helpful to be able to incorporate bracketing standards, e.g. for ion counter yield corrections on samples.

**Comment by bowring** [comment]
Preserve original flavor Tripoli behavior

**Reply by ryanickert** [reply]
Yeah I'm not sure how you balance this with being flexible (which I also want!  and a pony!).  The fact that I can just run tripoli on the lab computer, my laptop, and my desktop without having to worry about ensuring that there are a lot of figity settings to adjust on each is really great.  If there was some way to ensure that a vanilla, fresh-out-of-the-box ν-Tripoli had basic functionality similar to OF-Tripoli would be great.

**Comment by noahmclean** [comment]
Chris Coath: we should talk outliers -- flag possible outliers for user attention and possible rejection.

**Reply by noahmclean** [reply]
Noah: We should outlaw 2sigma automatic rejection, with 3 points off your geochemistry license.

**Reply by ryanickert** [reply]
Seconded!

**Comment by noahmclean** [comment]
Chris C. - online data saving for Tripoli so you can share data with partners.

**Comment by schoeneblair** [comment]
do beam interpolation? would be interesting to explore other algorithms compared to those implemented by mass spec company. Build in ways to deal with those uncertainties (weren't you working on this Noah?).

**Comment by schoeneblair** [comment]
I have more than once wished I could export all the data (corrected, outlier rejected, etc.) as a .csv, or into clipboard. I believe it only does means and stats now.

**Reply by bowring** [reply]
make an issue: https://github.com/bowring/Tripoli/issues. might be easy to do.

**Reply by bowring** [reply]
that's old tripoli

**Reply by ryanickert** [reply]
you can save a csv (or maybe tab delimted?) file with all the cycle-by-cycle data (rej are multiplied by -1) on origina flavor tripoli using "Control Panel", "Save work and export as Text"

**Reply by schoeneblair** [reply]
ah, thanks!

**Comment by noahmclean** [comment]
Chris Coath: Create a human-readable guide to the workflow.  Summary statistics including # of cycles rejected.

**Reply by ashcocc** [reply]
If you provide references to documentation with the full equations for each correction, then the reports can be more concise

**Comment by noahmclean** [comment]
Ashley C: have a way to input and flag primary and secondary standards.

**Comment by noahmclean** [comment]
Chris: Once you've established your data reduction protocol, press a button to save it so that it's reproducible.  Automate!

**Comment by noahmclean** [comment]
Ashley: we need the whole periodic table, including all the actinides.

**Reply by noahmclean** [reply]
Chris: what about polyatomic molecules?  Brave new world of collision cell MSs this might be ambiguous.

**Comment by mtmohr** [comment]
Saving customized API layouts- a user (or lab group) could set up custom layout templates for data visualizations/compilations (not necessarily time series). Ability to attach metadata to a tripoli output (name of Analyst, spike name and weight, etc).

**Comment by noahmclean** [comment]
Blair: deadtime analysis: calculate from standard data, record in log, uncertainty from deadtime?

**Comment by noahmclean** [comment]
Blair: having lots of options for corrections and user math are important.  It's important for these to be traceable.

**Reply by noahmclean** [reply]
Seconded by Graham.

**Reply by ChristopherDCoath** [reply]
Thirded by me.  My first ever GitHub post!

**Reply by grahamedwards** [reply]
Seconded indeed! I think if would be especially cool to not only have a wide range of options for pre-written/built-in correction algorithms, but also a way to add custom functions. For instance, I've tested "experimental" datum-by-datum corrections in IsotopX's user maths / user functions (definitely not algorithms that we would want to be built into tripoli!). Tweaking parts of the correction requires re-processing each .xlsx datafile through the mass-spec software. Would be awesome if all of these tweaks and "re-processes" could be done offline in Tripoli "user-maths" that could then be efficiently applied to other samples. 

This way, the mass spec software is just used to collect data and Tripoli (+ downstream softwares as needed) would be in charge of all the processing/corrections.

**Comment by grahamedwards** [comment]
Hierarchical data management? Once mass spec data is imported into Tripoli, it could have a user interface to sort and organize files into hierarchical groups as well as apply tags/flags (e.g. ~~standard~~ reference material, project name, user), much like a reference manager with folders and subfolders behind the scenes. 

This would be especially nice to centralize everything in tripoli rather than having to worry about messing up the file organization requirements of mass spec software. Import raw mass spec data into tripoli and it lives there forever. Especially straightforward if Tripoli can do all the corrections we could ever desire so we don't need to go back to the mass spec software ([I'm a broken record](https://github.com/CIRDLES/Tripoli/discussions/34#discussioncomment-2963372)).

**Reply by bowring** [reply]
Great ideas!  Any chance of getting an example output folder for your machine - plus the machine specs, method(s) used, etc?

**Reply by grahamedwards** [reply]
Thanks! Output folder would be a bit tricky for me to procure -- I'm between instruments right now (all computational post-doc). I was working on Terry Blackburn's IsotopX Phoenix (vintage 2016 software). Maybe one of the more modern Phoenix users could spare one?

**Reply by ryanickert** [reply]
Happy to supply any or all of the data I have from our new PhoeniX but it will likely look pretty similar to Noah's data.  Unless you want the really wacky stuff, like my single-peak 232Th method or my 235U-237Np oxide method.

**Comment by noahmclean** [comment]
Sean Gaynor: we'll need a way to match up the isotope names as input in mass spectrometer methods to species that Tripoli knows, e.g. to do a meaningful isobaric interference ratio.  For instance, 206Pb, Pb206, and 206 all mean the isotope 206Pb.  Likewise, 270, 238UO2, and 238U16O16O might all reference the same isotope beam.

**Reply by SeanPG36** [reply]
Thanks for adding this Noah, now officially have an account.

**Comment by noahmclean** [comment]
Sean Gaynor: Make a text note field that also goes into the Tripoli output.

**Reply by noahmclean** [reply]
@bowring: We can add this to the Tripoli output with a new otuput schema.

**Comment by krchamberlain** [comment]
2 ideas. 1) a workaround for recording beam intensity (206 or 205Pb) during a S54 Single collector run. One option suggested was to use General Editor with the single beam interpolations processed in Tripoli and beam intensity crudely monitored in Faradays.
2) output from U-Pb Redux for Pb-Pb isochron values and uncertainties, including correlation coefficient for 206/204 vs 207/204.

**Comment by krchamberlain** [comment]
3rd idea: is there any consideration for applying Tripoli to SIMS, in particular CAMECA SIMS?  I have a lot of experience with U-Pb data on CAMECA and would like to be able to generate different curve styles (linear, power law, exponential, etc) for the calibrations with standards.  The SHRIMP software has this built in, but CAMECA are generally inhouse programs.  I can communicate more if you are interested.

**Comment by bowring** [comment]
@krchamberlain - noted!
